File: examples/cpp/source/neural_networks/abs_layer_dense_batch.cpp
/* file: abs_layer_dense_batch.cpp */
/*******************************************************************************
* Copyright 2014-2019 Intel Corporation.
*
* This software and the related documents are Intel copyrighted  materials,  and
* your use of  them is  governed by the  express license  under which  they were
* provided to you (License).  Unless the License provides otherwise, you may not
* use, modify, copy, publish, distribute,  disclose or transmit this software or
* the related documents without Intel's prior written permission.
*
* This software and the related documents  are provided as  is,  with no express
* or implied  warranties,  other  than those  that are  expressly stated  in the
* License.
*******************************************************************************/

/*
!  Content:
!    C++ example of forward and backward absolute value (abs) layer usage
!
!
!******************************************************************************/

/**
 * <a name="DAAL-EXAMPLE-CPP-ABS_LAYER_BATCH"></a>
 * \example abs_layer_dense_batch.cpp
 */

#include "daal.h"
#include "service.h"

using namespace std;
using namespace daal;
using namespace daal::algorithms;
using namespace daal::algorithms::neural_networks::layers;
using namespace daal::data_management;
using namespace daal::services;

/* Input data set parameters */
string datasetName = "../data/batch/layer.csv";

int main()
{
    /* Read datasetFileName from a file and create a tensor to store input data */
    TensorPtr tensorData = readTensorFromCSV(datasetName);

    /* Create an algorithm to compute forward abs layer results using default method */
    abs::forward::Batch<> absLayerForward;

    /* Set input objects for the forward abs layer */
    absLayerForward.input.set(forward::data, tensorData);

    /* Compute forward abs layer results */
    absLayerForward.compute();

    /* Print the results of the forward abs layer */
    abs::forward::ResultPtr forwardResult = absLayerForward.getResult();
    printTensor(forwardResult->get(forward::value), "Forward abs layer result (first 5 rows):", 5);

    /* Get the size of forward abs layer output */
    const Collection<size_t> &gDims = forwardResult->get(forward::value)->getDimensions();
    TensorPtr tensorDataBack = TensorPtr(new HomogenTensor<>(gDims, Tensor::doAllocate, 0.01f));

    /* Create an algorithm to compute backward abs layer results using default method */
    abs::backward::Batch<> absLayerBackward;

    /* Set input objects for the backward abs layer */
    absLayerBackward.input.set(backward::inputGradient, tensorDataBack);
    absLayerBackward.input.set(backward::inputFromForward, forwardResult->get(forward::resultForBackward));

    /* Compute backward abs layer results */
    absLayerBackward.compute();

    /* Print the results of the backward abs layer */
    backward::ResultPtr backwardResult = absLayerBackward.getResult();
    printTensor(backwardResult->get(backward::gradient), "Backward abs layer result (first 5 rows):", 5);

    return 0;
}

File: examples/python/source/neural_networks/abs_layer_dense_batch.py
# file: abs_layer_dense_batch.py
#===============================================================================
# Copyright 2014-2019 Intel Corporation.
#
# This software and the related documents are Intel copyrighted  materials,  and
# your use of  them is  governed by the  express license  under which  they were
# provided to you (License).  Unless the License provides otherwise, you may not
# use, modify, copy, publish, distribute,  disclose or transmit this software or
# the related documents without Intel's prior written permission.
#
# This software and the related documents  are provided as  is,  with no express
# or implied  warranties,  other  than those  that are  expressly stated  in the
# License.
#===============================================================================

#
# !  Content:
# !    Python example of forward and backward absolute value (abs) layer usage
# !
# !
# !*****************************************************************************

#
## <a name="DAAL-EXAMPLE-PY-ABS_LAYER_BATCH"></a>
## \example abs_layer_dense_batch.py
#

import os
import sys

from daal.algorithms.neural_networks import layers
from daal.data_management import HomogenTensor, TensorIface

utils_folder = os.path.realpath(os.path.abspath(os.path.dirname(os.path.dirname(__file__))))
if utils_folder not in sys.path:
    sys.path.insert(0, utils_folder)
from utils import printTensor, readTensorFromCSV


# Input data set parameters
datasetName = os.path.join("..", "data", "batch", "layer.csv")

if __name__ == "__main__":

    # Read datasetFileName from a file and create a tensor to store input data
    tensorData = readTensorFromCSV(datasetName)

    # Create an algorithm to compute forward abs layer results using default method
    absLayerForward = layers.abs.forward.Batch()

    # Set input objects for the forward abs layer
    absLayerForward.input.setInput(layers.forward.data, tensorData)

    # Compute forward abs layer results
    forwardResult = absLayerForward.compute()

    # Print the results of the forward abs layer
    printTensor(forwardResult.getResult(layers.forward.value), "Forward abs layer result (first 5 rows):", 5)

    # Get the size of forward abs layer output
    gDims = forwardResult.getResult(layers.forward.value).getDimensions()
    tensorDataBack = HomogenTensor(gDims, TensorIface.doAllocate, 0.01)

    # Create an algorithm to compute backward abs layer results using default method
    absLayerBackward = layers.abs.backward.Batch()

    # Set input objects for the backward abs layer
    absLayerBackward.input.setInput(layers.backward.inputGradient, tensorDataBack)
    absLayerBackward.input.setInputLayerData(layers.backward.inputFromForward, forwardResult.getResultLayerData(layers.forward.resultForBackward))

    # Compute backward abs layer results
    backwardResult = absLayerBackward.compute()

    # Print the results of the backward abs layer
    printTensor(backwardResult.getResult(layers.backward.gradient), "Backward abs layer result (first 5 rows):", 5)
